### DISCUSSION TITLE
Can global procedures be editable in regular tiddlers edit view?

### DISCUSSION THREAD

**Discussion by twMat** [TiddlyWiki/TiddlyWiki5#8586] [Ideas]
Tiddlers tagged `$:/tags/Global` or `.../Macro` are iffy to edit for two reasons:

* they trigger refresh if edited "live"
* they are stored in *separate* tiddlers so you must navigate back and forth to where the procedure is invoked

Idea: Could there in the (regular) tiddlers edit mode, above the text field, be an editview template that *transcludes* all locally used procedure definitions in an editor"? The procedures etc are still externally stored, and global, but now editable very conveniently. Positioned above the text field, they'd even be positioned at basically at the same location as local procedures are defined!

ADDED: The idea would mitigate the "navigate back and forth" bit, not the refresh triggering. The circumvention for the latter could be to merely show a *copy* of the procedure code in the proposed editor, and a button to overwrite the content into the actual procedure tiddler. (I use this method in the [SideEditor](https://sideeditor.tiddlyspot.com/) for editing tiddlers tagged `$:/tags/Macro` (not yet updated to handle `/Global`))

Whatyathink?

**Comment by CodaCodr** [comment]
> they trigger refresh if edited "live"

Every keypress (among other things) triggers a refresh -- nothing to do with how your tiddler is tagged, or what it contains.

Your second point is kinda hard to follow. It amounts to saying "Tiddlers tagged X ... are stored in _separate_ tiddlers." :/

I think you're looking for a sandbox that will allow you to edit freely, save, the push another button that publishes the tiddler to the live wiki "above". Sounds like something innerwiki might do.

**Comment by twMat** [comment]
>Every keypress (among other things) triggers a refresh -- nothing to do with how your tiddler is tagged, or what it contains.

Oh, but so why are only procedures/macros problematic to edit live? Focus is lost, for example.

>Mat said:
>procedures are stored in separate tiddlers so you must navigate back and forth to where the procedure is invoked

What I mean is the problems that come when you author complex code in parallel with procedures. When the two parts are defined in the same tiddler, it is only a minor hassle to scroll up and down between them, but when procedures are externalized into separate tiddlers, it is significantly more annoying to modify and edit code.

>I think you're looking for a sandbox that will allow you to edit freely, save, the push another button that publishes the tiddler to the live wiki "above". Sounds like something innerwiki might do.

It doesn't have to go as far as a proper sandbox or an innerwiki; in the [SideEditor plugin](https://sideeditor.tiddlyspot.com/) the macro text is simply copied into a temporary tiddler for edit and then copy it back via a button click. But, yes, the main point is that the editing should be much simpler even if the procedure is in an external tiddler.

**Comment by CodaCodr** [comment]
Okay, I'm following much better now. Yes, the interminable loss of focus thing. I did create a workaround mostly targeting single line `$edit-text` controls but you're right about developing long-form content and code "in-wiki". I'm working with one right now, it's _really_ slow while editing, too.

> Oh, but so why are only procedures/macros problematic to edit live?

Actually, I thought about that a bit more. _Perhaps_ it's _my_ wikis behaving like that: @Telmiger's word counter _might_ be to blame -- not really sure. It certainly _works_ keypress-by-keypress.

But to answer your question, "why", I'm not sure. They're in draft mode and should not (I believe) affect anything else in the store (which is what TW should be referencing).

Aside: I also use a split-editor where a given long tiddler can be edited "twice" in the same view -- saves on the scrolling up and down quite a bit. You could do the same with a modified editTemplate.

**Comment by kookma** [comment]
> * they trigger refresh if edited "live"

I think the PageTemplate imports all macro/procedures no matter they are draft or not!
The import should correct with adding `!is[draft]`

**Reply by Jermolene** [reply]
The page template imports macros via this filter expression, which definitely does include `!is[draft]`:

https://github.com/TiddlyWiki/TiddlyWiki5/blob/master/core/wiki/config/GlobalImportFilter.tid

**Comment by twMat** [comment]
@kookma 

>I think the PageTemplate imports all macro/procedures no matter they are draft or not!
>The import should correct with adding !is[draft]

Do you mean when using the edit mode **preview** feature? (Because, surely, you don't have that problem in "pure" edit mode, right? @CodaCodr refers to Elmigers "word counter", which I'm not familiar with, but it does sound plausible that such a feature might count in-editor words as they are typed. But that is not the native situation in TW, I'd say.)

>The import should correct with adding !is[draft]

For edit mode *preview*, I can see how this would be a problem; should it show an actual live preview or should it not?

Note that my OP doesn't request to solve this particular aspect of the problem. I'm just stating that it *is* a problem when working with procedures etc. My OP is more about "geographical proximity" so to say. I'll edit it to make this clearer.

**Reply by kookma** [reply]
> Do you mean when using the edit mode **preview** feature? (Because, surely, you don't have that problem in "pure" edit mode, right?

Please ignore my comment, as the link above (given by Jeremy) the `is[draft]` is included. 

> Note that my OP doesn't request to solve this particular aspect of the problem. I'm just stating that it _is_ a problem when working with procedures etc. My OP is more about "geographical proximity" so to say. I'll edit it to make this clearer.

That's true.  May be before editing, removing the $:/tags/Macro helps. After finishing the edit, one can return the tag.
